codalab/lib/zip_util.py

```diff
@@ -51,43 +40,48 @@ def strip_archive_ext(path):
+    Unpack the archive |source| to |dest_path|.
     Note: |source| can be a file handle or a path.
+    |ext| contains the extension of the archive.
     """
-    # Unpack to a temporary location.
-    # TODO: guard against zip bombs.  Put a maximum limit and enforce it here.
-    # In the future, we probably don't want to be unpacking things all over the place.
-    tmp_path = tempfile.mkdtemp('-zip_util.unpack')
-    if isinstance(source, basestring):
-        source_path = source
-        if source_path.endswith('tar.gz') or source_path.endswith('tgz'):
-            exitcode = subprocess.call(['tar', 'xfz', source_path, '-C', tmp_path])
-        elif source_path.endswith('tar.bz2'):
-            exitcode = subprocess.call(['tar', 'xfj', source_path, '-C', tmp_path])
-        elif source_path.endswith('zip'):
-            exitcode = subprocess.call(['unzip', '-q', source_path, '-d', tmp_path])
-        elif source_path.endswith('.gz'):
-            with open(os.path.join(tmp_path, os.path.basename(strip_archive_ext(source_path))), 'wb') as f:
-                exitcode = subprocess.call(['gunzip', '-q', '-c', source_path], stdout=f)
-        else:
-            raise UsageError('Not an archive: %s' % source_path)
-        if exitcode != 0:
-            raise UsageError('Error unpacking %s' % source_path)
-    else:
-        # File handle, stream the contents!
-        source_handle = source
-        proc = subprocess.Popen(['tar', 'xfz', '-', '-C', tmp_path], stdin=subprocess.PIPE)
-        file_util.copy(source_handle, proc.stdin, print_status='Downloading and unpacking to %s' % tmp_path)
-        proc.stdin.close()
-        proc.wait()
+    if ext != '.zip':
+        close_source = False
+        try:
+            if isinstance(source, basestring):
+                source = open(source, 'rb')
+                close_source = True
 
-    # Move files into the right place.
-    # If archive only contains one path, then use that.
-    files = [f for f in os.listdir(tmp_path) if not ignore_file(f)]
-    if len(files) == 1:
-        path_util.rename(os.path.join(tmp_path, files[0]), dest_path)
-        path_util.remove(tmp_path)
+            if ext == '.tar.gz' or ext == '.tgz':
+                un_tar_directory(source, dest_path, 'gz')
+            elif ext == '.tar.bz2':
+                un_tar_directory(source, dest_path, 'bz2')
+            elif ext == '.gz':
+                with open(dest_path, 'wb') as f:
+                    shutil.copyfileobj(un_gzip_stream(source), f)
+            else:
+                raise UsageError('Not an archive.')
+        except (tarfile.TarError, IOError):
+            raise UsageError('Invalid archive upload.')
+        finally:
+            if close_source:
+                source.close()
     else:
-        path_util.rename(tmp_path, dest_path)
+        delete_source = False
+        try:
+            # unzip doesn't accept input from standard input, so we have to save
```

Comment:
That's unfortunate.  The CLI and worker both use tar, right?  I think zip is only used when people upload from the website.